core/examples/restore_checkpoint_delta.rs

```diff
@@ -0,0 +1,213 @@
+    let start = Instant::now();
+    restorer.start_to_restore(sm2.clone());
+    while !restorer.progress().is_completed() {
+        std::thread::sleep(Duration::from_secs(1));
+    }
+    println!("restoration completed in {:?}", start.elapsed());
+    let restored_root = restorer.restored_state_root(sm2.clone()).delta_root;
+    println!("restored root: {:?}", restored_root);
+    assert_eq!(restored_root, checkpoint_root);
+
+    // validate restoration
+    let epoch_id = SnapshotAndEpochIdRef::new(&checkpoint, None);
+    let state = sm2.get_state_no_commit(epoch_id)?.unwrap();
+
+    assert_eq!(
+        state.get("123".as_bytes())?,
+        Some(vec![1, 2, 3].into_boxed_slice())
+    );
+    assert_eq!(
+        state.get("124".as_bytes())?,
+        Some(vec![1, 2, 4].into_boxed_slice())
+    );
+    assert_eq!(
+        state.get("234".as_bytes())?,
+        Some(vec![2, 3, 4].into_boxed_slice())
+    );
+    assert_eq!(
+        state.get("235".as_bytes())?,
+        Some(vec![2, 3, 5].into_boxed_slice())
+    );
+
+    Ok(())
+}
+
+fn new_state_manager(db_dir: &str) -> Result<Arc<StateManager>, Error> {
+    create_dir_all(db_dir)?;
+
+    let db_config = db::db_config(
+        Path::new(db_dir),
+        Some(128),
+        db::DatabaseCompactionProfile::default(),
+        NUM_COLUMNS.clone(),
+        false,
+    );
+    let db = db::open_database(db_dir, &db_config)?;
+
+    Ok(Arc::new(StateManager::new(
+        db,
+        StorageConfiguration::default(),
+    )))
+}
+
+fn initialize_genesis(
+    manager: &StateManager,
+) -> Result<(H256, MerkleHash), Error> {
+    let mut state = manager.get_state_for_genesis_write();
+
+    state.set("123".as_bytes(), vec![1, 2, 3].into_boxed_slice())?;
+    state.set("124".as_bytes(), vec![1, 2, 4].into_boxed_slice())?;
+
+    let root = state.compute_state_root()?;
+    println!("genesis root: {:?}", root.state_root.delta_root);
+
+    let genesis_hash = H256::from_str(
+        "fa4e44bc69cca4cb2ae88a8fd452826faab9e8764e7eed934feede46c98962fa",
+    )
+    .unwrap();
+    state.commit(genesis_hash.clone())?;
+
+    Ok((genesis_hash, root.state_root.delta_root))
+}
+
+fn prepare_checkpoint(
+    manager: &StateManager, parent: &H256, accounts: usize,
+) -> Result<(H256, MerkleHash), Error> {
+    let checkpoint = H256::random();
+
+    let mut state = manager
+        .get_state_for_next_epoch(SnapshotAndEpochIdRef::new(parent, None))?
+        .unwrap();
+
+    state.set("234".as_bytes(), vec![2, 3, 4].into_boxed_slice())?;
+    state.set("235".as_bytes(), vec![2, 3, 5].into_boxed_slice())?;
+    state.compute_state_root()?;
+    state.commit(checkpoint)?;
+
+    println!("begin to add {} accounts for checkpoint ...", accounts);
+    let mut root = H256::zero();
+    let start = Instant::now();
+    let mut pending = accounts;
+    while pending > 0 {
+        let n = min(100_000, pending);
+        let start2 = Instant::now();
+        root = add_accounts(manager, checkpoint, n);
```

Comment:
1. can not commit multiple times on the same epoch_id. Change checkpoint to root.
2. When pending == 1, add_accounts(manager, checkpoint, n)